### PR TITLE
Fix e-mail URI escaping recipients

### LINKED ISSUE #12010: iOS Email composing crashes App
### Description

I am sending an email using the Email.Default.ComposeAsync() method

this works fine on Android but crashes on iOS with the exception below
```
2022-12-09 19:19:10.850 ClientUI[11511:2319672]    at System.Uri..ctor(String uriString, UriKind uriKind)
   at Foundation.NSUrl.op_Implicit(NSUrl url)
   at Microsoft.Maui.ApplicationModel.Communication.EmailImplementation.ComposeWithUrl(EmailMessage message)
2022-12-09 19:19:10.850 ClientUI[11511:2319672]    at Microsoft.Maui.ApplicationModel.Communication.EmailImplementation.PlatformComposeAsync(EmailMessage message)
   at Microsoft.Maui.ApplicationModel.Communication.EmailImplementation.ComposeAsync(EmailMessage message)
```

here is the code that send the email
```
if( Email.Default.IsComposeSupported )
        {
          string username = name;
          if( string.IsNullOrEmpty( username ) )
          {
            username = "Unknown";
          }

          string subject = $"Wi-Fi Messenger Logs for client <{username}>";
          string body = $"Attached is the zipped up log files for client <{username}>";
          string targetName = Path.Combine( TARGET_PATH,
                                            MakeValidFileName( $"{username}_{DateTime.UtcNow:yyyy-MM-dd_hh-mm-ss}.zip" ) );

          using FileStream targetFile = File.OpenWrite( targetName );
          //Emails require the zip file be in an actual file
          outputMemStream.WriteTo( targetFile );
          targetFile.Close();

          EmailMessage message = new()
          {
            Subject = subject,
            Body = body,
            BodyFormat = EmailBodyFormat.PlainText,
            Attachments = new List<EmailAttachment>() { new EmailAttachment( targetName ) },
            To = new List<string>() { recipient.Trim() }
          };

          await Email.Default.ComposeAsync( message );
```

I have added this to my info.plist
```
<key>LSApplicationQueriesSchemes</key>
<array>
  <string>mailto</string>
</array>
```

### Steps to Reproduce

1. Create a maui App
2. try to send an Email using the Email Api on a physical iOS device
3. Observe the crash

### Link to public reproduction project repository

N/A

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS

### Affected platform versions

iOS 16.1.2

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Cheesebaron**
### Description of Change

Escaping the recipient e-mail addresses, leads to the character @ being escaped and creates an invalid URL.

So `john@myhost.com` becomes `john%40myhost.com` which isn't containing a valid e-mail.

On iOS this fails when translating the URL to NSUrl.

### Issues Fixed

Fixes #12010

### CODE REVIEW

**Comment by Cheesebaron** [comment]
Hold your horses, just reading RFC 2368, there are even more issues with this method for creating the URI. I guess no one ever tested it?

Everything is delimited using `?` this is wrong. Only the first parameter should be formatted like that.

**Comment by Cheesebaron** [comment]
I've changed the implementation, with inspiration from https://codereview.stackexchange.com/a/129600, to properly generate a RFC 2368 mailto URL.

So previously there were several issues.

- `To`, `CC`, `BCC` where all URL escaped, which is not correct, recipients should not be escaped, only subject and body should be escaped
- The `To` recipients were delimited incorrectly for the format chosen

This has now been corrected and proper URLs are emitted. Perhaps someone should add some implementation to the e-mail tests ;)

**Comment by Cheesebaron** [comment]
Also find it a bit weird why [`MailMessage`](https://learn.microsoft.com/en-us/dotnet/api/system.net.mail.mailmessage?view=net-7.0) is not used instead of rolling your own? Any particular reason for this?

**Comment by jfversluis** [comment]
@Cheesebaron just an FYI we didn't forget about this one! Matt had some off time and trying to figure out the history here if there was a good reason for it. Otherwise I think this is a great improvement. Thank you for that! I'll make sure we get to it shortly :)

**Comment by Cheesebaron** [comment]
> @Cheesebaron just an FYI we didn't forget about this one! Matt had some off time and trying to figure out the history here if there was a good reason for it. Otherwise I think this is a great improvement. Thank you for that! I'll make sure we get to it shortly :)

Yep no worries. I have a workaround in my App for this which works, so I am not in a rush to get this in.

**Comment by mattleibow** [comment]
The lack of MailMessage may have been a result of UWP... We can move to that if it will solve more issues that the effort to fix this now?

**Comment by mattleibow** [comment]
Also:

> Perhaps someone should add some implementation to the e-mail tests

Thanks for starting the work on the tests! ;)

File: src/Essentials/src/Email/Email.shared.cs [hunk 1]
```diff
@@ -60,31 +60,29 @@ public Task ComposeAsync(EmailMessage? message)
...
-				parts.Add("body=" + Uri.EscapeDataString(message!.Body));
-			if (!string.IsNullOrEmpty(message?.Subject))
-				parts.Add("subject=" + Uri.EscapeDataString(message!.Subject));
-			if (message?.Cc?.Count > 0)
-				parts.Add("cc=" + Uri.EscapeDataString(string.Join(",", message.Cc)));
-			if (message?.Bcc?.Count > 0)
-				parts.Add("bcc=" + Uri.EscapeDataString(string.Join(",", message.Bcc)));
+		static IEnumerable<string> Parameters(EmailMessage message)
+		{
+			if (message.To?.Any() == true)
```
Comment (mattleibow) [inline src/Essentials/src/Email/Email.shared.cs:68]: Using LINQ and Any() is not so performant. Maybe it is now, but we can also just use ?.Count > 0 which we know is fast.

Comment (hartez) [reply src/Essentials/src/Email/Email.shared.cs:68]: Nobody is composing emails in a tight loop, I think we can let this go and worry about the performance characteristics of this another day 🙂

**Comment by Cheesebaron** [comment]
> Thanks for starting the work on the tests! ;)

🙈

I was having trouble even building the solution, but got that to work yesterday. I can have a look at adding some of those. There are no tests at all for any of this e-mail stuff to begin with 😂

File: src/Essentials/src/Email/Email.shared.cs (hunk 1 above)
Comment (Cheesebaron) [reply src/Essentials/src/Email/Email.shared.cs:68]: Up to you whether you want me to use Count() > 0 or Any(). I have no strong feelings about this one.

Comment (Cheesebaron) [reply src/Essentials/src/Email/Email.shared.cs]: Fixed. Can you have another look?

**Review by mattleibow — APPROVED** [review]

**Comment by mattleibow** [comment]
Amazing!

> There are no tests at all for any of this e-mail stuff

No idea what you mean! 😉 All I see are wonderful tests in this PR 😉 Perfect example of a PR and tests to go along with it!

Thanks a big bunch!

**Comment by BioTurboNick** [comment]
Is there a way to work around this until the fix is released? I now realize that I've been seeing this crash in AppCenter because people needing help weren't able to contact me...

**Comment by ederbond** [comment]
Hi @Cheesebaron , could you show us what's on your `Email.ComposeAsync(message)` on line 6 of your code above?
Who is that `Email`class?

**Comment by Cheesebaron** [comment]
> Hi @Cheesebaron , could you show us what's on your `Email.ComposeAsync(message)` on line 6 of your code above? Who is that `Email`class?

That is just MAUI Essentials

**Comment by ederbond** [comment]
Backport it to .NET 7.

**Pull request merged**